libs/vkd3d/memory.c

```diff
@@ -659,9 +756,303 @@ void vkd3d_memory_allocator_cleanup(struct vkd3d_memory_allocator *allocator, st
 
     vkd3d_free(allocator->chunks);
     vkd3d_va_map_cleanup(&allocator->va_map);
+    vkd3d_memory_allocator_cleanup_clear_queue(allocator, device);
     pthread_mutex_destroy(&allocator->mutex);
 }
 
+static HRESULT vkd3d_memory_allocator_flush_clears_locked(struct vkd3d_memory_allocator *allocator,
+        struct d3d12_device *device)
+{
+    const VkPipelineStageFlags vk_stage_mask = VK_PIPELINE_STAGE_ALL_COMMANDS_BIT;
+    struct vkd3d_memory_clear_queue *clear_queue = &allocator->clear_queue;
+    const struct vkd3d_vk_device_procs *vk_procs = &device->vk_procs;
+    VkTimelineSemaphoreSubmitInfoKHR timeline_info;
+    struct vkd3d_queue *queue, *internal_queue;
+    VkCommandBufferBeginInfo begin_info;
+    VkMemoryBarrier memory_barrier;
+    VkCommandBuffer vk_cmd_buffer;
+    VkCommandPool vk_cmd_pool;
+    VkSubmitInfo submit_info;
+    uint32_t queue_mask;
+    VkQueue vk_queue;
+    VkResult vr;
+    size_t i;
+
+    if (!clear_queue->allocations_count)
+        return S_OK;
+
+    /* Record commands late so that we can simply remove allocations from
+     * the queue if they got freed before the clear commands got dispatched,
+     * rather than rewriting the command buffer or dispatching the clear */
+    internal_queue = device->queues[VKD3D_QUEUE_FAMILY_INTERNAL_COMPUTE];
+    vk_cmd_buffer = clear_queue->vk_command_buffers[clear_queue->command_buffer_index];
+    vk_cmd_pool = clear_queue->vk_command_pools[clear_queue->command_buffer_index];
+
+    if ((vr = VK_CALL(vkResetCommandPool(device->vk_device, vk_cmd_pool, 0))))
+    {
+        ERR("Failed to reset command pool, vr %d.\n", vr);
+        return hresult_from_vk_result(vr);
+    }
+
+    begin_info.sType = VK_STRUCTURE_TYPE_COMMAND_BUFFER_BEGIN_INFO;
+    begin_info.pNext = NULL;
+    begin_info.flags = VK_COMMAND_BUFFER_USAGE_ONE_TIME_SUBMIT_BIT;
+    begin_info.pInheritanceInfo = NULL;
+
+    if ((vr = VK_CALL(vkBeginCommandBuffer(vk_cmd_buffer, &begin_info))) < 0)
+    {
+        ERR("Failed to begin command buffer, vr %d.\n", vr);
+        return hresult_from_vk_result(vr);
+    }
+
+    for (i = 0; i < clear_queue->allocations_count; i++)
+    {
+        const struct vkd3d_memory_allocation *allocation = clear_queue->allocations[i];
+
+        VK_CALL(vkCmdFillBuffer(vk_cmd_buffer, allocation->resource.vk_buffer,
+                allocation->offset, allocation->resource.size, 0));
+    }
+
+    /* Buffers are CONCURRENT due to D3D12 API restrictions,
+     * so we don't need to bother with ownership transfers */
+    memory_barrier.sType = VK_STRUCTURE_TYPE_MEMORY_BARRIER;
+    memory_barrier.pNext = NULL;
+    memory_barrier.srcAccessMask = VK_ACCESS_TRANSFER_WRITE_BIT;
+    memory_barrier.dstAccessMask = 0;
+
+    VK_CALL(vkCmdPipelineBarrier(vk_cmd_buffer,
+            VK_PIPELINE_STAGE_TRANSFER_BIT,
+            VK_PIPELINE_STAGE_BOTTOM_OF_PIPE_BIT,
```

Comment:
A memory barrier here is moot since we're using semaphores to synchronize the access with other queues.

Reply:
Don't we at least need to make sure that the writes from vkCmdFillBuffer actually becomes available?

Reply:
Signaling a semaphore makes all device writes available. Waiting for semaphore makes all device memory available.

Reply:
Actually we do still need the barrier if this maps to a user queue since we will not wait on the timeline semaphore on that particular queue.

Reply:
Changed the logic here so that we wait on all queues except the internal compute one if it is unique.